### PR TITLE
Linear color math, attempt three

### PR DESCRIPTION
**Pull request by jpark37**
### Description
This is yet another attempt to make GPU sRGB formats work in OBS. This attempt is far more backwards-compatible than past attempts.

Any texture that is created as GS_RGBA, GS_BGRA, or GS_BGRX will support both sRGB and non-sRGB usages. There are new sRGB variants of texture setting functions, and a global setting to activate sRGB writes. On the D3D side, this means deferring output target binding until draw time, which doesn't seem to be a big deal.

This way, existing sources and filters do not need to change, and just operate as if sRGB support didn't exist. The small commits in this PR show gradual migration is possible. Another benefit is that I'm not performing as many unnecessary sRGB conversions as the last PR attempt since it's okay to pass through a nonlinear value as long as you aren't trying to do linear math on it.

**EDIT (2021-01-21): All existing sources and filters have been updated anyway.**

### Motivation and Context
- "Better"-looking resizing.
- More accurate color filter settings.

### How Has This Been Tested?
I've tested all four sRGB usages commits on Windows (D3D) and Mac (GL). with a checkboard texture consisting of 0/0/0 and 188/188/188 squares. When bilinear-filtered by half in both dimensions with gamma conversion on both texture lookup and render target write, the result should be 137/137/137.

The performance impact is somehow much less than when I had measured it many months ago on Intel HD Graphics 530. (Performance appears identical on AMD.) This is possibly a side effect of cumulative OBS GPU optimizations? OS/driver upgrades? I don't really have a satisfying hard answer for this, but I'll take it!

Test Scenario 1, UNORM to SRGB Differences:
597x647 Item Render -> 597x647 Canvas
1024x742 Item Render -> [Area] -> 597x422 Canvas
644x478 Item Render -> [Area (Upscale)] -> 1323x1080 Canvas

Test Scenario 2, UNORM to SRGB Differences:
1920x1080 Image -> 1920x1080 Canvas
1920x1080 Image -> 1920x1080 Intermediate -> [scale filter (bilinear)] -> 960x540 Canvas
1920x1080 Image -> 1920x1080 Intermediate -> [bilinear] -> 960x540 Canvas
1920x1080 Image -> [default] -> 960x540 Canvas
1920x1080 Image -> [final downsample (bilinear)] -> 960x540 Canvas

Test Scenario 1, Intel HD Graphics 530, SetStablePowerState
TotalFrame: 4385 us -> 4564 us
 OutputFrame: 3691 us -> 3873 us
  Video: 3672 us -> 3853 us
   MainTexture: 1998 us -> 2065 us
   OutputTexture: [no downscale, no draw]
   ConvertTexture: 851 us -> 954 us
   StageOutput: 782 us -> 780 us
 RenderDisplays: 684 us -> 681 us

Test Scenario 2, Intel HD Graphics 530, SetStablePowerState
TotalFrame: 5026 us -> 5361 us
 OutputFrame: 4305 us -> 4679 us
  Video: 4282 us -> 4658 us
   MainTexture: 3504 us -> 3689 us
   OutputTexture: 338 us -> 522 us
   ConvertTexture: 213 us -> 219 us
   StageOutput: 211 us -> 212 us
 RenderDisplays: 718 us -> 678 us

**EDIT (2020-12-28): New testing**

Having come back to this PR after about a year of inactivity, I've created a new test plan. I'm glossing over the details for brevity, but the following checklist should provide the gist.

- [x] Test 3D LUT on GL for regression (new GLSL implementation for raw texture loads).
- [x] Test rescale via source setting operates in linear space.
- [x] Test rescale via scale filter operates in linear space.
- [x] Test Apply LUT filter does **not** operate in linear space.
- [x] Test Color Correction filter does **not** operate in linear space.
- [x] Test video source with deinterlacing enabled to exercise new GL 2D load shader function.
- [x] Test video downsample operates in linear space.
- [x] Test game capture for regressions to exercise new UNORM-specific color formats.
- [x] Test device loss recovery.
- [x] Test shared texture fallback still works.
- [x] Record video with NVENC to verify NV12 still works.
- [x] Record video with AMF just in case.
- [x] Test all changes on Windows D3D.
- [x] Test all changes on Windows GL.
- [x] Test all changes on Linux.
- [x] Test all changes on Mac.

**EDIT (2021-01-21): Combine source/transition/filter PRs #4082 and #4083**

Sources:
- [x] decklink\decklink-source.cpp (async)
- [x] image-source\color-source.c
- [x] image-source\image-source.c
- [x] image-source\obs-slideshow.c
- [x] Screen Capture (XSHM)
- [x] Window Capture (Xcomposite)
- [x] linux-v4l2\v4l2-input.c (async)
- [x] mac-avcapture\av-capture.mm (async)
- [x] mac-capture\mac-display-capture.m
- [x] mac-capture\mac-window-capture.m (async)
- [x] obs-ffmpeg\obs-ffmpeg-source.c (async)
- [x] obs-text\gdiplus\obs-text.cpp
- [x] text-freetype2\text-freetype2.c
- [x] vlc-video\vlc-video-source.c (async)
- [x] win-capture\duplicator-monitor-capture.c
- [x] win-capture\game-capture.c
- [x] win-capture\monitor-capture.c
- [x] win-capture\window-capture.c
- [x] win-dshow\win-dshow.cpp (async)

Transitions:
- [x] obs-transitions\transition-fade.c
- [x] obs-transitions\transition-fade-to-color.c
- [x] obs-transitions\transition-luma-wipe.c
- [x] obs-transitions\transition-slide.c
- [x] obs-transitions\transition-swipe.c

Filters:
- [x] Apply LUT
- [x] Chroma Key
- [x] Color Correction
- [x] Color Key
- [x] Image Mask/Blend
- [x] Luma Key
- [x] Render Delay
- [x] Scaling/Aspect Ratio
- [x] Sharpen

### Types of changes
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by kkartaltepe** [comment]
It would be nice to note the extension requirement for OpenGL added(newly required, though it looks like it is part of the existing loader config)

**Comment by jpark37** [comment]
I'm not sure exactly how widespread EXT_texture_sRGB_decode is, but I would imagine it's near universal on 3.3 hardware because all Valve source games require it to run on Linux, and they were working with IHVs to fix their OpenGL drivers for a good while.

I think I can cleanly fix the OBS_ALLOW_DIRECT_RENDERING issue by providing an override toggle to redirect texture binds to be sRGB for updated filters. If I make that the only method of altering texture binds, I can also kill off some of the other added functions although it might be nice to keep them around for convenience.

**Comment by jpark37** [comment]
This PR is really close. The last bug that I'm chasing down is wrong video output color when using Ubuntu WSL. I'm able to run Linux RenderDoc on WSL, and it's showing the textures look good all the way through on the GPU side, so I'm hoping this is just a bug in the WSL driver stack when staging textures for read. I'll have my Linux PC to test on soon, so I'll be able to verify then.

**Comment by jpark37** [comment]
Upon further inspection, this final detail is actually quite annoying: texelFetch is specified to ignore  SKIP_DECODE_EXT. I think Mesa is doing the right thing, but my GPU drivers on Windows/Mac are doing the wrong thing. Ugh.

I can think of plenty of suboptimal solutions, but nothing I actually want to implement. Going to mull this over for a while.

**Comment by jpark37** [comment]
For GL textures read by Load calls, we could add internal bool uniforms (obs_srgb_encode_[name]) to the generated shader to signal whether or not the texelFetch SRGB conversion needs to be undone. These would be updated at texture bind time.

There's an annoying amount of plumbing to figure out, but I don't see a cleaner approach. It's also a bit of a performance hit for GL platforms, but I don't think it'll be horrendous.

It would look something like this.

Current:

    float4 sample = texelFetch(foo, ...);

Proposed (auto-generated):

    static float3 obs_srgb_encode(float3 sample)
    {
        // (channel < 0.0031308) ? (channel * 12.92) : ((1.055 * pow(channel, 1.0 / 2.4)) - 0.055);
        ...
    }
    ...
    float4 sample = texelFetch(foo, ...);
    if (obs_glsl_compile && obs_srgb_encode_foo)
        sample.rgb = obs_srgb_encode(sample.rgb);

**Comment by jpark37** [comment]
This PR now depends on #2284 to reduce 3D texture changes.

Unfortunately, it appears proper format aliasing was not added in OpenGL until 4.2 via glBindImageTexture. To work around SKIP_DECODE_EXT not working with texelFetch, I have added a new "extension method" to the Texture object named LoadRaw, which works around the problem by re-encoding the values with extra shader math. HLSL still runs full speed because the non-sRGB SRV is bound, and the symbol LoadRaw is replaced with Load.

The only known issue is to scrub through the shaders, and figure out which Load calls need to be replaced with LoadRaw. Then testing...

**Comment by jp9000** [comment]
Alrighty, merged #2284.

**Comment by jpark37** [comment]
@jp9000 On second thought, while "LoadRaw" works, I feel like it's too messy and awkward. I have two alternatives.

- Have two GL textures internally for formats with sRGB alternatives, and perform fake aliasing by copying between the two as necessary. Granted, this is more expensive than some shader math to undo the implicit sRGB decode, but the plumbing should be less invasive: no more shenanigans involving ARB_framebuffer_sRGB and EXT_texture_sRGB_decode, and no shader changes/hints are required.
- Add ARB_texture_view support if the GPU driver supports it for real aliasing. The GPUs below have GL 4.3 support. We'd still need the fake aliasing path for Mac (no GL 4.3 support), and older cards. Mac will go back to full speed once Metal support is added.

GPUs that should support ARB_texture_view:

- AMD Radeon HD 5000 Series and newer. (2009)
- Intel HD Graphics in Intel Haswell processors and newer. (2013)
- Nvidia GeForce 400 series and newer. (2010)

What do you think?

**Comment by jp9000** [comment]
Er, wait, one comment back, didn't we recently upgrade to 3.3?  Something tells me a couple people are still going to be using Intels older than what you mentioned with Linux.  I remember someone complained when we upgraded from 3.2 to 3.3.  Although to be fair, the people on those hardware tend to be crazy and I'm not sure if we should care that much, because as time progresses it should be okay to move up and deprecate older things sometimes.  So I don't really care that much.  But backward compatibility is always nice too.  But I'm getting sidetracked.

So is this problem strictly GL?  It sounds strictly like a GL problem.

Wouldn't using the extension when it's available be better for performance?  It sounds like option 2 also contains option 1 as a backup, so I'm sort of confused as to why you'd have option 1 in the first place.  Maybe I'm misreading it.  What's the downside of 2 if it already contains option 1 I mean?  More invasive plumbing (aka messier code)?

I feel like performance is probably of greatest importance, but if it's *just* GL, then I feel like I'd be willing to make a *minor* sacrifice if it means really simple/clean code too.  If the cost is just extra VRAM memory or something then I feel like that's a fine sacrifice.  If the cost is something like a percentage of performance then I'm worried.

Basically I'll probably be okay with whichever.  If you want to make some sacrifice, I feel find with it just as long as it's only GL and the impact is insignificant.  I feel like I trust your judgment here because you tend to know my preference prioritization (that priority being: performance first when possible, then clean/simple design and clean/simple code next best thing).

**Comment by jpark37** [comment]
Yeah, we just moved "up" to 3.3. I think the 3.2 -> 3.3 causing issues was a red herring though. Nobody actually got hosed by moving to 3.3 from my knowledge since there weren't any 3.2 GPUs that couldn't do 3.3.

Yes, this is a GL problem. D3D11 supports format aliasing out of the box.

Option 2 is better for performance, but it'll take me longer to stand up, and might increase future maintenance from being an additional code path. I'll probably end up doing it though.

**Comment by jpark37** [comment]
Actually, I can avoid both LoadRaw and texture copy solutions by just emulating texelFetch with textureLod and  textureSize. Unlike texelFetch, textureLod should honor SKIP_DECODE_EXT.

```
vec4 obs_load_raw_3d(sampler3D s, ivec4 p_lod)
{
    int lod = p_lod.w;
    vec3 size = textureSize(s, lod);
    vec3 p = vec3(p_lod.xyz) / size;
    vec4 color = textureLod(s, p, lod);
    return color;
}

vec4 obs_load_raw_2d(sampler2D s, ivec3 p_lod)
{
    int lod = p_lod.z;
    vec2 size = textureSize(s, lod);
    vec2 p = vec2(p_lod.xy) / size;
    vec4 color = textureLod(s, p, lod);
    return color;
}
```

**Comment by jpark37** [comment]
The PR is finally in good state, and I have completed testing. The only thing keeping this PR in draft status is a dependency on PR #3919.

There is still more work to be done to update the rest of the sources and filters, but there shouldn't be any regressions by leaving them as-is. I have only updated the image source, and the scale filter in this PR. Some filters like color correction should probably be deprecated and replaced entirely to ensure we don't break existing scenes; most (all?) adjustments should be computed in a linear space, and the way we handle brightness by adding/subtracting gray is weird to me.

**Comment by jpark37** [comment]
Draft status removed. Let me know if anything looks questionable.

**Comment by trwnh** [comment]
On Linux (using obs-studio-git), color correction now no longer works and has no properties as of this commit/PR.

**Comment by jpark37** [comment]
> On Linux (using obs-studio-git), color correction now no longer works and has no properties as of this commit/PR.

Should be addressed by #4122.
